src/Features/Core/Portable/NavigateTo/RoslynNavigateToItem.cs

```diff
@@ -0,0 +1,251 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Runtime.Serialization;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Navigation;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.NavigateTo
+{
+    /// <summary>
+    /// Data about a navigate to match.  Only intended for use by C# and VB.  Carries enough rich information to
+    /// rehydrate everything needed quickly on either the host or remote side.
+    /// </summary>
+    [DataContract]
+    internal readonly struct RoslynNavigateToItem
+    {
+        [DataMember(Order = 0)]
+        public readonly bool IsStale;
+
+        [DataMember(Order = 1)]
+        public readonly DocumentId DocumentId;
+
+        [DataMember(Order = 2)]
+        public readonly string CombinedProjectName;
+
+        [DataMember(Order = 3)]
+        public readonly DeclaredSymbolInfo DeclaredSymbolInfo;
+
+        [DataMember(Order = 4)]
+        public readonly string Kind;
+
+        [DataMember(Order = 5)]
+        public readonly NavigateToMatchKind MatchKind;
+
+        [DataMember(Order = 6)]
+        public readonly bool IsCaseSensitive;
+
+        [DataMember(Order = 7)]
+        public readonly ImmutableArray<TextSpan> NameMatchSpans;
+
+        public RoslynNavigateToItem(
+            bool isStale,
+            DocumentId documentId,
+            string combinedProjectName,
+            DeclaredSymbolInfo declaredSymbolInfo,
+            string kind,
+            NavigateToMatchKind matchKind,
+            bool isCaseSensitive,
+            ImmutableArray<TextSpan> nameMatchSpans)
+        {
+            IsStale = isStale;
+            DocumentId = documentId;
+            CombinedProjectName = combinedProjectName;
+            DeclaredSymbolInfo = declaredSymbolInfo;
+            Kind = kind;
+            MatchKind = matchKind;
+            IsCaseSensitive = isCaseSensitive;
+            NameMatchSpans = nameMatchSpans;
+        }
+
+        public bool TryCreateSearchResult(Solution solution, [NotNullWhen(true)] out INavigateToSearchResult? result)
+        {
+            var document = solution.GetDocument(DocumentId);
+            if (document == null)
+            {
+                var project = solution.GetProject(DocumentId.ProjectId);
+                if (project != null)
+                    document = project.TryGetSourceGeneratedDocumentForAlreadyGeneratedId(DocumentId);
+            }
+            Contract.ThrowIfTrue(document == null && !IsStale, "We should always be able to map back a non-stale result to the solution");
+
+            if (document == null)
```

Comment:
⚠️ `Contract.ThrowIfTrue` is not a debug assertion. This condition will always be false.

Reply:
the contract call is more specific.  it checks null && IsStale.